### PR TITLE
SISRP-21098 Retrofit class calendar integration to EDO DB for non-legacy terms

### PR DESCRIPTION
**Pull request by paulkerschen**
https://jira.berkeley.edu/browse/SISRP-21098

Previously on our show, #5517 created the database classes.

To keep things organized, a fair amount of functionality is moved from Calendar::Preprocessor into Calendar::QueuedEntry and Calendar::Schedule. Since this feature is mostly tested at the integration level, we modify existing tests for Calendar::Preprocessor rather than trying to split them out. The big integration testext example is modified likewise.

### CODE REVIEW

File: app/models/calendar/queued_entry.rb [hunk 1]
```diff
@@ -9,6 +9,101 @@ class QueuedEntry < ActiveRecord::Base
...
+        year: row['term_yr'],
+        term_cd: row['term_cd'],
+        ccn: row['course_cntl_num'],
+        multi_entry_cd: row['multi_entry_cd'].blank? ? '-' : row['multi_entry_cd']
+      ).first_or_initialize
+
+      entry.name = row['course_name']
+      entry.schedule = schedule
+
+      if row['building_name'].present?
```
Comment (johncrossman) [inline app/models/calendar/queued_entry.rb]: @paulkerschen We could lighten the lines below by making the line above:
 `if (building_name = row['building_name']).present?`
After all, summer calls for lightness.

File: app/models/calendar/queued_entry.rb [hunk 2]
```diff
@@ -9,6 +9,101 @@ class QueuedEntry < ActiveRecord::Base
...
+        term_cd: term_code[:term_cd],
+        ccn: row['section_id'],
+        # The name 'multi_entry_cd' is a carry-over from a legacy Oracle column. We use it to
+        # disambiguate multiple meeting entries for the same class section.
+        multi_entry_cd: row['meeting_num'].to_s
+      ).first_or_initialize
 
+      entry.name = row['section_display_name']
+      entry.schedule = schedule
+      entry.location = "#{row['location']}, UC Berkeley" if row['location'].present?
```
Comment (johncrossman) [inline app/models/calendar/queued_entry.rb]: @paulkerschen When experimenting with Google Maps, it seems that a suffix of `..., Berkeley, CA` would be more effective than `..., UC Berkeley`. For example:
* `Wheeler 150, Berkeley, CA` is better than `Wheeler 150, UC Berkeley`
* `2195 Hearst 330C, Berkeley, CA` is better
* `Genetics & Plant Bio 104, Berkeley, CA` is better
* Etc.
IMO, this feature would benefit from locations that Google can recognize.

File: app/models/calendar/schedule.rb [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        times
+      end
+
+      WEEKDAYS = %w(SU MO TU WE TH FR SA)
+
+      def advance_time(time_start, time, ampm_flag)
+        time_stripped = time.gsub(/^0/, '')
+        minute = time_stripped.slice!(-2, 2).to_i
+        hour = time_stripped.to_i
+        if hour < 12 &&  ampm_flag == 'P'
```
Comment (johncrossman) [inline app/models/calendar/schedule.rb]: @paulkerschen stray space before `ampm_flag`

File: app/models/calendar/preprocessor.rb [hunk 4]
```diff
@@ -11,123 +11,51 @@ def initialize
...
-          elsif rrule.blank?
-            logger.info "Blank recurrence rule, this will be a DELETE action"
-            entry.transaction_type = Calendar::QueuedEntry::DELETE_TRANSACTION
-          else
-            logger.info "This will be an UPDATE action"
-            entry.transaction_type = Calendar::QueuedEntry::UPDATE_TRANSACTION
+    def get_legacy_entries(entries)
+      CampusOracle::Calendar.get_all_courses.each do |course_row|
+        if (entry = Calendar::QueuedEntry.from_legacy_row course_row)
+          enrollments = CampusOracle::Calendar.get_whitelisted_students_in_course(@users, course_row['term_yr'], course_row['term_cd'], course_row['course_cntl_num'])
```
Comment (johncrossman) [inline app/models/calendar/preprocessor.rb]: @paulkerschen Two minor requests:

1. Can we take this opportunity to shorten the method name `get_whitelisted_students_in_course`  :)
1. A brief comment to explain purpose of this `get_legacy_entries(entries)` method

**Comment by johncrossman** [comment]
Looks good, @paulkerschen I've put a handful of non-urgent questions / comments. A quick glance from @raydavis might be a good idea.

**Comment by paulkerschen** [comment]
Thanks for review, @johncrossman; PR amended to address. I'd also be happy for @raydavis to glance over if he has the time.

**Comment by raydavis** [comment]
Thanks for the initial review, @johncrossman! And thank you not only for the retrofit, @paulkerschen , but also for making the old stuff easy to yank out in two months.

**Pull request merged**